server/chat-plugins/friends.ts

```diff
@@ -0,0 +1,753 @@
+/**
+ * Friends list plugin.
+ * Allows for adding and removing friends, as well as seeing their activity.
+ * Written by Mia-pi.
+ * @author mia-pi-git
+ */
+import * as Sqlite from 'better-sqlite3';
+import {Utils} from '../../lib/utils';
+import {FS} from '../../lib/fs';
+
+/** Max friends per user */
+const MAX_FRIENDS = 100;
+/** Max friend requests. */
+const MAX_REQUESTS = 6;
+/** `Map<transferTo, transferFrom>` */
+export const transferRequests: Map<string, string> = Chat.oldPlugins.friends?.transferRequests || new Map();
+
+const STATUS_COLORS: {[k: string]: string} = {
+	idle: '#ff7000',
+	online: '#009900',
+	busy: '#cc3838',
+};
+
+const STATUS_TITLES: {[k: string]: string} = {
+	online: 'Online',
+	idle: 'Idle',
+	busy: 'Busy',
+	offline: 'Offline',
+};
+
+export const Friends = new class {
+	/** `Map<oldID, newID> */
+	renames: Map<string, string>;
+	private readonly database: Sqlite.Database;
+	private readonly statements: {[k: string]: Sqlite.Statement};
+	constructor() {
+		this.database = new Sqlite(`${__dirname}/../../databases/friends.db`);
+		this.renames = new Map();
+		// remove this when sqlite tables are automatically loaded
+		// check if the tables exist - if not; create them
+		try {
+			this.getRenames();
+		} catch (e) {
+			if (e.message.includes('no such table:')) {
```

Comment:
Instead of doing this I would check if `friends.db` exists, and if not, create it with the current schema. Eventually this will have to be generalized to "if friends.db exists, but its schema version is different than the current, run transitions"

Reply:
This doesn't work because `better-sqlite3` creates the db file on creation of the class (eg `new require('better-sqlite3')('test.db')`` creates the file `test.db`) even if it's not there already.

Reply:
I suppose you could initialize the table before the rest of the database, but that seems messier than what's there now.

Reply:
[See `fileMustExist`](https://github.com/JoshuaWise/better-sqlite3/blob/master/docs/api.md#new-databasepath-options) (also, even if this doesn't work for you, just check before creating the database object?)

Reply:
Done.